File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/signature/KeyId.java
package uk.nhs.nhsx.core.signature;

import uk.nhs.nhsx.core.ValueType;

public class KeyId extends ValueType<KeyId> {

    private KeyId(String value) {
        super(value);
    }
    
    public static KeyId of(String idOrArn) {
        return new KeyId(kmsKeyId(idOrArn));
    }

    private static String kmsKeyId(String kmsKeyArnOrId) {
        if (kmsKeyArnOrId.startsWith("arn:aws:kms:")) {
            return kmsKeyArnOrId.substring(kmsKeyArnOrId.indexOf(":key/") + ":key/".length());
        }
        return kmsKeyArnOrId;
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/signature/KeyId.kt
package uk.nhs.nhsx.core.signature

import dev.forkhandles.values.StringValue

class KeyId private constructor(value: String) : StringValue(value) {
    companion object {
        fun of(idOrArn: String) = KeyId(kmsKeyId(idOrArn))

        private fun kmsKeyId(kmsKeyArnOrId: String) = if (kmsKeyArnOrId.startsWith("arn:aws:kms:")) {
            kmsKeyArnOrId.substring(kmsKeyArnOrId.indexOf(":key/") + ":key/".length)
        } else kmsKeyArnOrId
    }
}
